### PR TITLE
Modernise packaging and CI

### LINKED ISSUE #253: CI: Update actions
Several issues:

- There's a separate workflow for the `master` branch -- can this just be merged into the main action?
- Master and publish workflows still run on Py3.7
- Publish workflow runs tests -- presumably at the point of making a new release, we've already run tests?
- Publish workflow should use `build` package instead of running `setup.py`
- Publish workflow should probably use API token instead of PYPI_PASSWORD
- `black` workflow is just check, but could be switched to auto-commit
- Various actions need versions bumping:

> Node.js 12 actions are deprecated. For more information see: https://github.blog/changelog/2022-09-22-github-actions-all-actions-will-begin-running-on-node16-instead-of-node12/. Please update the following actions to use Node.js 16: actions/checkout, actions/setup-python, codecov/codecov-action, actions/checkout

### PR DESCRIPTION
**Pull request by ZedThree**
Fixes #253 

- Switch completely to `pyproject.toml`, remove `setup.cfg`, `setup.py`
- Various updates to CI, including:
  - auto-apply black
  - use new ["trusted publisher"](https://docs.pypi.org/trusted-publishers/) workflow for publishing

### CODE REVIEW

**Comment by dschwoerer** [comment]
ERROR: Could not find a version that satisfies the requirement animatplot-ng==0.4.2 (from versions: 0.4.3)
That was my mistake - I shouldn't have switched them to animatplot-ng

I am happy to remove those tests.
I am confused how xarray manages to prevent installing on python < 3.9 - I would assume the old versions did not include that at some point? Are they pulling in something, that requires a new python?

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -20,56 +20,56 @@ jobs:
     if: always()
     strategy:
       matrix:
-        python-version: [3.8, 3.9, '3.10', 3.x]
-        pip-packages:
-          - "setuptools pip pytest pytest-cov coverage boutdata xarray numpy>=1.16.0"
+        python-version: ["3.8", "3.9", "3.10", "3.x"]
```
Comment (dschwoerer) [inline .github/workflows/master.yml:23]:
```suggestion
        python-version: ["3.8", "3.9", "3.10", "3.11", "3.x"]
```

File: .github/workflows/master.yml [hunk 2]
```diff
@@ -20,56 +20,56 @@ jobs:
...
 
     runs-on: ubuntu-latest
     if: always()
     strategy:
       matrix:
-        python-version: [3.8]
+        python-version: ["3.8"]
         pip-packages:
-          - "setuptools pip pytest pytest-cov coverage boutdata==0.1.4 xarray==0.18.0 dask==2.10.0 numpy==1.18.0 natsort==5.5.0 matplotlib==3.1.1 animatplot==0.4.2 netcdf4==1.4.2 Pillow==6.1.0" # test with oldest supported version of packages. Note, using numpy==1.18.0 as a workaround because numpy==1.17.0 is not supported on Python-3.7, even though we should currently support numpy==1.17.0.
+          - "boutdata==0.1.4 xarray==0.18.0 dask==2.10.0 numpy==1.18.0 natsort==5.5.0 matplotlib==3.1.1 animatplot-ng==0.4.2 netcdf4==1.4.2 Pillow==6.1.0"
```
Comment (dschwoerer) [inline .github/workflows/master.yml:53]:
```suggestion
          - "boutdata==0.1.4 xarray==0.18.0 dask==2.10.0 numpy==1.18.0 natsort==5.5.0 matplotlib==3.1.1 animatplot==0.4.2 netcdf4==1.4.2 Pillow==6.1.0"
```

File: .github/workflows/pythonpackage.yml [hunk 3]
```diff
@@ -20,57 +20,56 @@ jobs:
...
 
     runs-on: ubuntu-latest
     if: always()
     strategy:
       matrix:
-        python-version: [3.8]
+        python-version: ["3.8"]
         pip-packages:
-          - "setuptools pip pytest pytest-cov coverage boutdata==0.1.4 xarray==0.18.0 dask==2.10.0 numpy==1.18.0 natsort==5.5.0 matplotlib==3.1.1 animatplot==0.4.2 netcdf4==1.4.2 Pillow==7.2.0" # test with oldest supported version of packages. Note, using numpy==1.18.0 as a workaround because numpy==1.17.0 is not supported on Python-3.7, even though we should currently support numpy==1.17.0.
+          - "boutdata==0.1.4 xarray==0.18.0 dask==2.10.0 numpy==1.18.0 natsort==5.5.0 matplotlib==3.1.1 animatplot-ng==0.4.2 netcdf4==1.4.2 Pillow==6.1.0"
```
Comment (dschwoerer) [inline .github/workflows/pythonpackage.yml:53]:
```suggestion
          - "boutdata==0.1.4 xarray==0.18.0 dask==2.10.0 numpy==1.18.0 natsort==5.5.0 matplotlib==3.1.1 animatplot==0.4.2 netcdf4==1.4.2 Pillow==6.1.0"
```

File: setup.py [hunk 4]
```diff
@@ -1,6 +0,0 @@
-#!/usr/bin/env python3
-
-from setuptools import setup
-
-if __name__ == "__main__":
-    setup(use_scm_version=True)
```
Comment (dschwoerer) [inline setup.py:6]: Not sure about removing this one - I thought it is still useful as a back up in some cases?
I forgot when, so might well be fine to remove it by now ...

**Review by dschwoerer — APPROVED** [review]

File: setup.py (hunk 4 above)
Comment (ZedThree) [reply setup.py:6]: I don't think this is ever needed any more. We can always restore it if someone screams

**Comment by ZedThree** [comment]
I think maybe an issue with the version of `netcdf4`?

```
Collecting netcdf4==1.4.2
  Downloading netCDF4-1.4.2.tar.gz (769 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 769.8/769.8 kB 53.3 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Installing backend dependencies: started
  Installing backend dependencies: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'error'
  error: subprocess-exited-with-error
  
  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      <string>:170: DeprecationWarning: The SafeConfigParser class has been renamed to ConfigParser in Python 3.2. This alias will be removed in future versions. Use ConfigParser directly instead.
      reading from setup.cfg...
      using nc-config ...
      Traceback (most recent call last):
        File "/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/opt/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 149, in prepare_metadata_for_build_wheel
          return hook(metadata_directory, config_settings)
        File "/tmp/pip-build-env-hmw2rbma/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 380, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "/tmp/pip-build-env-hmw2rbma/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 487, in run_setup
          super(_BuildMetaLegacyBackend,
        File "/tmp/pip-build-env-hmw2rbma/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 338, in run_setup
          exec(code, locals())
        File "<string>", line 457, in <module>
      ModuleNotFoundError: No module named 'numpy'
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed
```

**Comment by johnomotani** [comment]
> Is `pytest-min-versions` still a useful test overall?

I think it's useful in principle, but I'd be happy to bump the versions up a bit. I think it's mostly been useful for when xarray introduces some new functionality/syntax - if you change something that uses a new feature it might break xBOUT for users with an older xarray. Even then it's fine to force them to upgrade, but you need to know to bump the minimum required version required for xarray, etc.

**Comment by johnomotani** [comment]
I think everything except `xarray` has a pretty stable API, and is not that likely to be called directly anyway from xBOUT. I'd be happy to do a 'min versions test' that just installs xarray-0.18.0, and lets everything else have whatever version it likes. Does that seems easier to get working?

The 'min versions' test has saved me a few times from using something that was only available in the very latest xarray, so it would be nice to keep in some form if possible.

**Comment by ZedThree** [comment]
Sure, happy to do that. Just to be clear, the other tests will only install xarray 2022.9.0, so they will also catch features only in more recent versions too.

**Pull request merged**